File: validate-ukeire.py
#!/usr/bin/env python3
"""
Comprehensive ukeire validation — computes reference output for all test hands.
Outputs JSON for comparison with validate-ukeire.js.

Usage: python3 validate-ukeire.py > ukeire-reference.json
"""
import json
from mahjong.shanten import Shanten
from mahjong.tile import TilesConverter

_S = Shanten()

def s34(t): return _S.calculate_shanten(t)

ALL_34_INFO = (
    [('m', v) for v in range(1, 10)] +
    [('p', v) for v in range(1, 10)] +
    [('s', v) for v in range(1, 10)] +
    [('z', v) for v in range(1, 8)]
)

def h(**kw):
    return TilesConverter.string_to_34_array(**kw)

def key(suit, value):
    return suit + str(value)

def compute_waits(t34):
    # Adding 1 tile must produce a valid count (≤ 14 for standard hand)
    if sum(t34) >= 14:
        return None  # skip — would produce 15-tile hand
    waits = []
    for j, (s, v) in enumerate(ALL_34_INFO):
        remaining = 4 - t34[j]
        if remaining <= 0: continue
        t = t34[:]
        t[j] += 1
        try:
            if s34(t) == -1:
                waits.append({'tile': key(s,v), 'remaining': remaining})
        except ValueError:
            pass
    return waits

def compute_ukeire(t34_orig):
    current = s34(t34_orig)
    results = []
    seen = set()
    for i, (suit, value) in enumerate(ALL_34_INFO):
        if t34_orig[i] == 0: continue
        if i in seen: continue
        seen.add(i)

        effective = []
        best_after = current
        for j, (ds, dv) in enumerate(ALL_34_INFO):
            remaining = 4 - t34_orig[j]
            if remaining <= 0: continue
            t_new = t34_orig[:]
            t_new[i] -= 1
            t_new[j] += 1
            sh = s34(t_new)
            if sh < current:
                effective.append({'tile': key(ds,dv), 'remaining': remaining})
            if sh < best_after:
                best_after = sh

        results.append({
            'discard': key(suit, value),
            'shantenAfter': best_after,
            'kinds': len(effective),
            'total': sum(e['remaining'] for e in effective),
            'effective': [e['tile'] for e in effective],
        })
    return sorted(results, key=lambda r: (r['shantenAfter'], -r['total']))

# ── All test hands ────────────────────────────────────────────────────────────
HANDS = [
    # ── riichi.wiki canonical: tenpai (shanten=0) ────────────────────────────
    {'id': 'rw-07', 'desc': 'riichi.wiki case 7: 333m567p12789s55z (tenpai)',
     't34': h(man='333', pin='567', sou='12789', honors='55')},
    {'id': 'rw-14', 'desc': 'riichi.wiki case 14: 199m99p55s223355z (chiitoi tenpai)',
     't34': h(man='199', pin='99', sou='55', honors='223355')},

    # ── riichi.wiki: complete (-1) ────────────────────────────────────────────
    {'id': 'rw-01', 'desc': 'riichi.wiki case 1: 258m258p258s1235z (6-shanten)',
     't34': h(man='258', pin='258', sou='258', honors='1235')},

    # ── MahjongRepository shanten extended tests ──────────────────────────────
    {'id': 'ext-01', 'desc': '111234567s + 11p + 567m (complete)',
     't34': h(sou='111234567', pin='11', man='567')},
    {'id': 'ext-02', 'desc': '111345677s + 11m + 567p (tenpai)',
     't34': h(sou='111345677', man='11', pin='567')},
    {'id': 'ext-03', 'desc': '11122245679999s (tenpai)',
     't34': h(sou='11122245679999')},
    {'id': 'ext-04', 'desc': '114477s + 114477p + 76m (chiitoi tenpai)',
     't34': h(sou='114477', pin='114477', man='76')},
    {'id': 'ext-05', 'desc': '114477s + 114479p + 76m (1-shanten chiitoi)',
     't34': h(sou='114477', pin='114479', man='76')},

    # ── Ukeire test hands ────────────────────────────────────────────────────
    {'id': 'uk-01', 'desc': 'shanpon 123456789m + 1155p',
     't34': h(man='123456789', pin='1155')},
    {'id': 'uk-02', 'desc': 'single wait 123456789m + 1255p',
     't34': h(man='123456789', pin='1255')},
    {'id': 'uk-03', 'desc': 'chiitoi tenpai 1122334455m + 66p + 7s',
     't34': h(man='1122334455', pin='66', sou='7')},
    {'id': 'uk-04', 'desc': 'kokushi tenpai 19s + 19p + 19m + 1234567z',
     't34': h(sou='19', pin='19', man='19', honors='1234567')},
    {'id': 'uk-05', 'desc': '1-shanten 111345677s + 56m + 33p',
     't34': h(sou='111345677', man='56', pin='33')},
    {'id': 'uk-06', 'desc': '1-shanten 123456789m + 23p + 56s',
     't34': h(man='123456789', pin='23', sou='56')},
    {'id': 'uk-07', 'desc': '1-shanten H1: 123456789m + 235p + 1z',
     't34': h(man='123456789', pin='235', honors='1')},
    {'id': 'uk-08', 'desc': 'H5: 456789m + 2355p + 789s',
     't34': h(man='456789', pin='2355', sou='789')},

    # ── Yaku test hands (complete hands) ─────────────────────────────────────
    {'id': 'yaku-01', 'desc': 'tanyao: 234567s + 234567m + 22p',
     't34': h(sou='234567', man='234567', pin='22')},
    {'id': 'yaku-02', 'desc': 'pinfu: 123456s + 123456m + 55p',
     't34': h(sou='123456', man='123456', pin='55')},
    {'id': 'yaku-03', 'desc': 'chinitsu: 11234567677889m',
     't34': h(man='11234567677889')},
    {'id': 'yaku-04', 'desc': 'honitsu: 123455667m + 11122z',
     't34': h(man='123455667', honors='11122')},
    {'id': 'yaku-05', 'desc': 'iipeikou: 112233s + 333m + 12344p',
     't34': h(sou='112233', man='333', pin='12344')},

    # ── Extra edge cases ──────────────────────────────────────────────────────
    {'id': 'edge-01', 'desc': '2-shanten: 4566677s + 1367p + 8m + 12z',
     't34': h(sou='4566677', pin='1367', man='8', honors='12')},
    {'id': 'edge-02', 'desc': 'ittsu tenpai: 123456789m + 123p + 22z (13 tiles)',
     't34': h(man='123456789', pin='123', honors='22')},
    {'id': 'edge-03', 'desc': 'sanshoku tenpai: 123s + 123p + 12399m + 33z',
     't34': h(sou='123', pin='123', man='12399', honors='33')},
]

output = []
for hand in HANDS:
    t34 = hand['t34']
    sh = s34(t34)
    entry = {'id': hand['id'], 'desc': hand['desc'], 'shanten': sh}
    if sh == 0:
        entry['waits'] = compute_waits(t34)
    elif sh > 0:
        entry['ukeire'] = compute_ukeire(t34)[:8]  # top 8 discards
    output.append(entry)

print(json.dumps(output, ensure_ascii=False, indent=2))

File: validate-ukeire.js
/**
 * Compares our ukeire.js output against the Python MahjongRepository reference.
 * Reads /tmp/ukeire-reference.json produced by validate-ukeire.py.
 *
 * Run: node validate-ukeire.js
 */
import { readFileSync } from 'fs';
import { computeUkeire, computeWaits, analyzeEfficiency } from './src/utils/mahjong/ukeire.js';
import { computeShanten } from './src/utils/mahjong/shanten.js';

// ── Helpers ───────────────────────────────────────────────────────────────────
function fromStr(suit, digits) {
  return [...String(digits)].map(d => ({ suit, value: parseInt(d, 10) }));
}
function tileKey(t) { return t.suit + t.value; }

// Re-build the same hands as validate-ukeire.py
const HANDS_JS = {
  'rw-07':    [...fromStr('m','333'), ...fromStr('p','567'), ...fromStr('s','12789'), ...fromStr('z','55')],
  'rw-14':    [...fromStr('m','199'), ...fromStr('p','99'),  ...fromStr('s','55'),    ...fromStr('z','223355')],
  'rw-01':    [...fromStr('m','258'), ...fromStr('p','258'), ...fromStr('s','258'),   ...fromStr('z','1235')],
  'ext-01':   [...fromStr('s','111234567'), ...fromStr('p','11'), ...fromStr('m','567')],
  'ext-02':   [...fromStr('s','111345677'), ...fromStr('m','11'), ...fromStr('p','567')],
  'ext-03':   [...fromStr('s','11122245679999')],
  'ext-04':   [...fromStr('s','114477'), ...fromStr('p','114477'), ...fromStr('m','76')],
  'ext-05':   [...fromStr('s','114477'), ...fromStr('p','114479'), ...fromStr('m','76')],
  'uk-01':    [...fromStr('m','123456789'), ...fromStr('p','1155')],
  'uk-02':    [...fromStr('m','123456789'), ...fromStr('p','1255')],
  'uk-03':    [...fromStr('m','1122334455'), ...fromStr('p','66'), ...fromStr('s','7')],
  'uk-04':    [...fromStr('s','19'), ...fromStr('p','19'), ...fromStr('m','19'), ...fromStr('z','1234567')],
  'uk-05':    [...fromStr('s','111345677'), ...fromStr('m','56'), ...fromStr('p','33')],
  'uk-06':    [...fromStr('m','123456789'), ...fromStr('p','23'), ...fromStr('s','56')],
  'uk-07':    [...fromStr('m','123456789'), ...fromStr('p','235'), ...fromStr('z','1')],
  'uk-08':    [...fromStr('m','456789'), ...fromStr('p','2355'), ...fromStr('s','789')],
  'yaku-01':  [...fromStr('s','234567'), ...fromStr('m','234567'), ...fromStr('p','22')],
  'yaku-02':  [...fromStr('s','123456'), ...fromStr('m','123456'), ...fromStr('p','55')],
  'yaku-03':  [...fromStr('m','11234567677889')],
  'yaku-04':  [...fromStr('m','123455667'), ...fromStr('z','11122')],
  'yaku-05':  [...fromStr('s','112233'), ...fromStr('m','333'), ...fromStr('p','12344')],
  'edge-01':  [...fromStr('s','4566677'), ...fromStr('p','1367'), ...fromStr('m','8'), ...fromStr('z','12')],
  'edge-02':  [...fromStr('m','123456789'), ...fromStr('p','123'), ...fromStr('z','22')],
  'edge-03':  [...fromStr('s','123'), ...fromStr('p','123'), ...fromStr('m','12399'), ...fromStr('z','33')],
};

// ── Load reference ────────────────────────────────────────────────────────────
const reference = JSON.parse(readFileSync('/tmp/ukeire-reference.json', 'utf8'));

// ── Compare ───────────────────────────────────────────────────────────────────
let pass = 0, fail = 0, skip = 0;
const failures = [];

for (const ref of reference) {
  const tiles = HANDS_JS[ref.id];
  if (!tiles) { skip++; continue; }

  const shanten = computeShanten(tiles, 0);

  // 1. Shanten must match
  if (shanten !== ref.shanten) {
    fail++;
    failures.push(`[${ref.id}] shanten: JS=${shanten} PY=${ref.shanten}  (${ref.desc})`);
    continue;
  }
  pass++;

  // 2a. Tenpai: compare waits
  if (ref.shanten === 0 && ref.waits !== null && ref.waits !== undefined) {
    const jsWaits  = computeWaits(tiles);
    const pyKeys   = new Set(ref.waits.map(w => w.tile));
    const jsKeys   = new Set(jsWaits.map(w => tileKey(w.tile)));

    // Wait tiles must match exactly
    const missingInJS = [...pyKeys].filter(k => !jsKeys.has(k));
    const extraInJS   = [...jsKeys].filter(k => !pyKeys.has(k));

    if (missingInJS.length || extraInJS.length) {
      fail++;
      failures.push(`[${ref.id}] waits mismatch:\n  PY=${[...pyKeys].sort().join(',')}  JS=${[...jsKeys].sort().join(',')}\n  (${ref.desc})`);
    } else {
      // Remaining counts must match
      let countMismatch = false;
      for (const pyW of ref.waits) {
        const jsW = jsWaits.find(w => tileKey(w.tile) === pyW.tile);
        if (!jsW || jsW.remaining !== pyW.remaining) {
          countMismatch = true;
          failures.push(`[${ref.id}] wait ${pyW.tile}: PY remaining=${pyW.remaining} JS=${jsW?.remaining}`);
          fail++;
          break;
        }
      }
      if (!countMismatch) { pass++; console.log(`  ✓ [${ref.id}] waits match (${ref.waits.length}种)  ${ref.desc}`); }
    }
    continue;
  }

  // 2b. Non-tenpai: compare GOOD discards only (shantenAfter ≤ original shanten)
  // Bad discards (shantenAfter > original) always return [] in the Python replacement
  // method but have effective tiles in JS sequential method — this is a known
  // methodological difference, not a bug. Tenhou 牌理 uses the JS approach.
  if (ref.shanten > 0 && ref.ukeire) {
    const jsUkeire  = computeUkeire(tiles);
    const goodPy    = ref.ukeire.filter(e => e.shantenAfter <= ref.shanten);

    if (goodPy.length === 0) {
      console.log(`  ○ [${ref.id}] no good discards in PY output (skip)  ${ref.desc}`);
      skip++; continue;
    }

    let allMatch = true;
    for (const pyEntry of goodPy) {
      // Skip entries where Python returns 0 effective tiles (bad discard per replacement method)
      if (pyEntry.effective.length === 0) continue;

      const jsEntry = jsUkeire.find(e => tileKey(e.discardTile) === pyEntry.discard);
      if (!jsEntry) {
        failures.push(`[${ref.id}] discard ${pyEntry.discard} missing in JS  (${ref.desc})`);
        fail++; allMatch = false; break;
      }

      const pyEffSet = new Set(pyEntry.effective);
      const jsEffSet = new Set(jsEntry.effectiveTiles.map(e => tileKey(e.tile)));
      const missing  = [...pyEffSet].filter(k => !jsEffSet.has(k));
      const extra    = [...jsEffSet].filter(k => !pyEffSet.has(k));

      if (missing.length || extra.length) {
        failures.push(`[${ref.id}] discard=${pyEntry.discard}:\n  PY=[${[...pyEffSet].sort()}]\n  JS=[${[...jsEffSet].sort()}]\n  (${ref.desc})`);
        fail++; allMatch = false; break;
      }
      if (jsEntry.totalCount !== pyEntry.total) {
        failures.push(`[${ref.id}] discard=${pyEntry.discard} total: PY=${pyEntry.total} JS=${jsEntry.totalCount}`);
        fail++; allMatch = false; break;
      }
    }
    if (allMatch) { pass++; console.log(`  ✓ [${ref.id}] ukeire match (${goodPy.length} good discards)  ${ref.desc}`); }
    continue;
  }

  // Complete or skip-waits hand
  console.log(`  ○ [${ref.id}] shanten=${ref.shanten} (skipped: complete or 14-tile tenpai)  ${ref.desc}`);
  skip++;
}

console.log(`\n${'─'.repeat(70)}`);
console.log(`  ${pass} passed  ${fail} failed  ${skip} skipped`);
if (fail === 0) console.log('  ✅ All compared results match Python reference!');
if (failures.length) {
  console.log('\nFailures:');
  failures.forEach(f => console.log('  ✗ ' + f));
}
console.log(`${'─'.repeat(70)}\n`);
if (fail > 0) process.exitCode = 1;
